pages/DSNP/Identity.md

```diff
@@ -26,6 +26,22 @@ the ability to delegate permission to others to announce content on the user's b
 * Announcements from a delegate MUST be able to be verified as to which delegate made the specific announcement.
 * Delegation revocation MUST NOT be retroactive.
 
+## Operations
+
+Each of the these MUST have verifiable approval of the acting
+principal(s) via signature proof or other system. For example, an
+implementation might use the public key of an asymmetric key pair as a
+control key, and provide a proof for each operation by producing a
+cryptographic signature of the inputs and some nonce value.
+
+| Operation | Optional? | Principal(s) | Inputs | Outputs/State Changes |
+|---------- |---------- |------------- |------- |-------------- |
+| Create DSNP Identifier | no | User | None | New DSNP Identifier connected to User's Control Key |
+| Create or Update Delegation | no | User AND Delegate | User's DSNP Identifier, Delegate's DSNP Identifier, Set of Announcement Permissions | Proof of Delegation |
+| Revoke Delegation | no | User OR Delegate | User's DSNP Identifier, Delegate's DSNP Identifier | Proof of Delegation Revocation |
+| Retire DSNP Identifier | no | User | User's DSNP Identifier | Allowed removal of all data sent from User's DSNP Identifier, Allowed removal of all Control Keys from DSNP Identifier, Allowed removal of all User's Delegations |
```

Comment:
So it looks like retiring an identifier is not actually described in the existing text of Identity.md, nor is it mentioned in the DSNP white paper. If it is a spec operation we should define it, which probably requires its own DIP and a spec version change.

Assuming we do define it, can we be more prescriptive here? If a DSNP Identifier is retired, and then there is a subsequent operation that attempts to use it, can we say it should not be allowed? (This is different from retaining data in storage, which is an implementation detail.)

Reply:
Hmm.... Perhaps it is an optional spec piece? You are right about the data. That is a very implementation specific.

I would be ok with optional or removing this and making it just a part of the implementation spec.

Reply:
I don't think this should be optional. Arguments in favor, in no particular order:

1. We talk and have written about the ability to delete/retire many times. Leaving it out of the specification implies the right to be forgotten is optional or even orthogonal to DSNP, which to me doesn't make sense, especially since we're about giving users control over their social identity. "[those] who can destroy a thing, control a thing", or so some Herbert dude once wrote.

2. If we specify Tombstones why would we not support retiring identifiers?

3. Why eliminate the 'D' from CRUD just because it's identifier? What purpose does it serve to specify creation, updates, reads, but deny deletion (retire)?

4. Specifications often evolve well beyond white papers, so I don't consider that a strong argument against retiring identifiers.

A counterargument occurred to me that we don't specify any other operations regarding legal compliance, however, this is really the only portion of the specification that really applies, and leaving it out would explicitly ignore such laws. 

> Assuming we do define it, can we be more prescriptive here? If a DSNP Identifier is retired, and then there is a subsequent operation that attempts to use it, can we say it should not be allowed?

I think this is the best route, specify what should happen when the various operations include a retired identifier.

Reply:
@shannonwells I agree with your points. I think "Retire Identifier" should be a required (and well specified) operation. Since that is a spec change I will propose a new DIP.

Reply:
DIP-220 has been created for the ID retirement topic.